src/xclim/indices/run_length.py

```diff
@@ -125,6 +131,43 @@ def resample_and_rl(
     return out
 
 
+def _smallest_uint(da, dim):
+    for dtype in [np.uint8, np.uint16, np.uint32, np.uint64]:
+        if np.iinfo(dtype).max > da[dim].size:
+            return dtype
+    return np.uint64
+
+
+# Specifying `one` allows in-place multiplication *=
+@njit
+def _cumsum_reset_np(arr, index, one):
+    """100110111 -> 100120123"""
+    # run the cumsum and prod backwards or forward
+    it = range(1, arr.shape[-1]) if index == "last" else range(arr.shape[-1] - 2, -1, -1)
+    for i in it:
+        # this works because we assume to have 1's and 0's
+        arr[..., i] *= arr[..., i - it.step] + one
+    return arr
+
+
+def _cumsum_reset_xr(da, dim, index, reset_on_zero):
+    """100110111 -> 100120123"""
+    # `index="first"` case: Algorithm is applied on inverted array and output is inverted back
+    if index == "first":
+        da = da[{dim: slice(None, None, -1)}]
+
+    # Example: da == 100110111 -> cs_s == 100120123
+    cs = da.cumsum(dim=dim)  # cumulative sum  e.g. 111233456
+    cond = da == 0 if reset_on_zero else da.isnull()  # reset condition
+    cs2 = cs.where(cond)  # keep only numbers at positions of zeroes e.g. N11NN3NNN (default)
+    cs2[{dim: 0}] = 0  # put a zero in front e.g. 011NN3NNN
+    cs2 = cs2.ffill(dim=dim)  # e.g. 011113333
+    out = cs - cs2
+    if index == "first":
+        out = out[{dim: slice(None, None, -1)}]
+    return out
+
+
```

Comment:
Maybe add corner tests for those core functions? I understand they are implicitly tested by many tests, but I think it'd be good to add tests for weird inputs, ie 0-length arrays, arrays with only NaNs, infs, etc. Stuff that would be not be exercised by other tests and that could provide safeguards when we try to optimize further.

Reply:
> ie 0-length arrays

I'm not sure why we would be interested in this? It's not a test I've ever seen in our testing suite at least.

> only NaNs

For rle, we expect boolean arrays. Even if you had nans in your data, typically da> thresh will get rid of nans and transform them to False. Currently, all-nan will just give all 0's, no runs. I can modify the return statement at the end of rle so that nans are reinserted at the end:
`return out.where(da)`

Reply:
I'll let you judge what kind of edge cases are useful and which are not, but I've seen xclim functions trip up on 0-length arrays resulting from all NaNs that were dropped (e.g. statistical fits), so I thought maybe this could happen here as well. 

Not sure I measure well the implications of this choice, so do what you think is best.